File: scripts/crypto-interop/verify_jobdek.mjs
/**
 * Interop lock for the D-PRV-7 job-dek delivery (THREAT-MODEL §4.2).
 *
 * Direction 1 (this script, Node = server role): generates a worker P-256
 * keypair + a DEK, seals the parcel EXACTLY like
 * server/api/security/vault/job-dek.post.js does (same primitives — keep in
 * sync), self-checks the opening, then writes vector-jobdek-node.json for
 * the Python side.
 *
 * Direction 2 (second pass): when vector-jobdek-py.json exists (written by
 * verify_jobdek.py, Python = server role), this script opens the
 * Python-sealed parcel with the worker private key — proving Node opens
 * Python parcels.
 *
 * Chain: node verify_jobdek.mjs → python verify_jobdek.py → node verify_jobdek.mjs
 *
 * Wire contract (do NOT deviate):
 *  - public keys: base64 of an uncompressed X9.62 point (65 bytes, 0x04);
 *  - shared = raw ECDH secret (32 bytes, x coordinate);
 *  - transport key = HKDF-SHA256(ikm=shared, salt=32 explicit zero bytes,
 *    info="nest2d-job-dek-v1", length 32);
 *  - parcel = base64( nonce(12) || ciphertext || tag(16) ),
 *    AES-256-GCM(transportKey, DEK, AAD = UTF-8 bytes of the job slug).
 */
import crypto from 'node:crypto'
import { existsSync, readFileSync, writeFileSync } from 'node:fs'

const HERE = new URL('.', import.meta.url)
const NODE_VECTOR = new URL('./vector-jobdek-node.json', import.meta.url)
const PY_VECTOR = new URL('./vector-jobdek-py.json', import.meta.url)

const HKDF_INFO = 'nest2d-job-dek-v1'
const HKDF_SALT = Buffer.alloc(32) // explicit zeros — part of the contract

// Same fallback as job-dek.post.js: some Node builds lack the secp256r1 alias.
function createP256Ecdh() {
    try {
        return crypto.createECDH('secp256r1')
    } catch {
        return crypto.createECDH('prime256v1')
    }
}

function sealParcel({ workerPubRaw, dek, jobSlug }) {
    const server = createP256Ecdh()
    server.generateKeys()
    const shared = server.computeSecret(workerPubRaw)
    const transportKey = Buffer.from(crypto.hkdfSync('sha256', shared, HKDF_SALT, HKDF_INFO, 32))
    const nonce = crypto.randomBytes(12)
    const cipher = crypto.createCipheriv('aes-256-gcm', transportKey, nonce, { authTagLength: 16 })
    cipher.setAAD(Buffer.from(jobSlug, 'utf8'))
    const parcel = Buffer.concat([nonce, cipher.update(dek), cipher.final(), cipher.getAuthTag()])
    shared.fill(0)
    transportKey.fill(0)
    return { serverPub: server.getPublicKey(), parcel }
}

function openParcel({ workerPrivRaw, serverPubRaw, parcelRaw, jobSlug }) {
    const worker = createP256Ecdh()
    worker.setPrivateKey(workerPrivRaw)
    const shared = worker.computeSecret(serverPubRaw)
    const transportKey = Buffer.from(crypto.hkdfSync('sha256', shared, HKDF_SALT, HKDF_INFO, 32))
    const nonce = parcelRaw.subarray(0, 12)
    const tag = parcelRaw.subarray(parcelRaw.length - 16)
    const ct = parcelRaw.subarray(12, parcelRaw.length - 16)
    const decipher = crypto.createDecipheriv('aes-256-gcm', transportKey, nonce, { authTagLength: 16 })
    decipher.setAAD(Buffer.from(jobSlug, 'utf8'))
    decipher.setAuthTag(tag)
    const dek = Buffer.concat([decipher.update(ct), decipher.final()])
    shared.fill(0)
    transportKey.fill(0)
    return dek
}

// --- Pass 1: Node seals (server role), self-checks, writes the vector. ---
if (!existsSync(PY_VECTOR)) {
    const jobSlug = 'interop-job-dek'
    const dek = crypto.randomBytes(32)
    const worker = createP256Ecdh()
    worker.generateKeys()
    const workerPrivRaw = worker.getPrivateKey()
    const workerPubRaw = worker.getPublicKey()
    if (workerPubRaw.length !== 65 || workerPubRaw[0] !== 0x04) {
        throw new Error('worker public key is not an uncompressed X9.62 point')
    }

    const { serverPub, parcel } = sealParcel({ workerPubRaw, dek, jobSlug })

    // Node↔Node sanity: the same worker key must open the parcel.
    const opened = openParcel({ workerPrivRaw, serverPubRaw: serverPub, parcelRaw: parcel, jobSlug })
    if (!opened.equals(dek)) throw new Error('Node self-check failed: parcel does not open')
    console.log('✓ Node seals and opens its own job-dek parcel (self-check)')

    writeFileSync(NODE_VECTOR, JSON.stringify({
        job_slug: jobSlug,
        dek: dek.toString('base64'),
        dek_fingerprint: crypto.createHash('sha256').update(dek).digest('hex'),
        worker_priv: workerPrivRaw.toString('base64'),
        worker_pub: workerPubRaw.toString('base64'),
        server_pub: serverPub.toString('base64'),
        parcel: parcel.toString('base64'),
    }, null, 2))
    console.log('✓ vector-jobdek-node.json written — now run verify_jobdek.py')
} else {
    // --- Pass 2: open the Python-sealed parcel (worker role). ---
    const nodeVector = JSON.parse(readFileSync(NODE_VECTOR, 'utf8'))
    const pyVector = JSON.parse(readFileSync(PY_VECTOR, 'utf8'))
    const dek = openParcel({
        workerPrivRaw: Buffer.from(nodeVector.worker_priv, 'base64'),
        serverPubRaw: Buffer.from(pyVector.server_pub, 'base64'),
        parcelRaw: Buffer.from(pyVector.parcel, 'base64'),
        jobSlug: pyVector.job_slug,
    })
    if (dek.toString('base64') !== pyVector.dek) {
        throw new Error('Python-sealed parcel opened to the WRONG dek')
    }
    console.log('✓ Node opens the Python-sealed job-dek parcel (HKDF + AES-GCM aligned)')
    console.log('\nJob-dek interop OK in both directions.')
}

File: scripts/crypto-interop/verify_jobdek.py
"""
Interop lock for the D-PRV-7 job-dek delivery (THREAT-MODEL §4.2).

Python = worker role, using the REAL unwrap path of worker_common.crypto
(`get_dek` with a stubbed HTTP transport) to open the parcel sealed by
scripts/crypto-interop/verify_jobdek.mjs (Node = server role). Then plays
the server role itself and writes vector-jobdek-py.json for the Node side
to open.

Run after verify_jobdek.mjs (first pass):
    python scripts/crypto-interop/verify_jobdek.py
Then re-run verify_jobdek.mjs to close the reverse direction.
"""
import base64
import hashlib
import json
import os
import sys

sys.path.insert(0, os.path.join(os.path.dirname(__file__), "..", "..", "workers", "common"))

from cryptography.hazmat.primitives import hashes, serialization  # noqa: E402
from cryptography.hazmat.primitives.asymmetric import ec  # noqa: E402
from cryptography.hazmat.primitives.ciphers.aead import AESGCM  # noqa: E402
from cryptography.hazmat.primitives.kdf.hkdf import HKDF  # noqa: E402

import worker_common.crypto as worker_crypto  # noqa: E402

HERE = os.path.dirname(__file__)
NODE_VECTOR = os.path.join(HERE, "vector-jobdek-node.json")
PY_VECTOR = os.path.join(HERE, "vector-jobdek-py.json")

_TRANSPORT_INFO = b"nest2d-job-dek-v1"
_TRANSPORT_SALT = b"\x00" * 32  # explicit zeros — part of the contract


def main() -> None:
    with open(NODE_VECTOR, encoding="utf-8") as fh:
        vector = json.load(fh)

    job_slug = vector["job_slug"]
    dek = base64.b64decode(vector["dek"])
    worker_priv_raw = base64.b64decode(vector["worker_priv"])
    worker_pub = base64.b64decode(vector["worker_pub"])

    # Rebuild the worker private key from its raw 32-byte scalar and lock the
    # point encoding agreement (derived pub must equal the Node-emitted pub).
    priv = ec.derive_private_key(
        int.from_bytes(worker_priv_raw, "big"), ec.SECP256R1()
    )
    derived_pub = priv.public_key().public_bytes(
        serialization.Encoding.X962, serialization.PublicFormat.UncompressedPoint
    )
    assert derived_pub == worker_pub, "P-256 point encoding mismatch Node vs Python"

    # --- Direction 1: open the Node-sealed parcel through the REAL worker
    # code path (get_dek), with only the HTTP transport stubbed. ---
    worker_crypto._job_priv_key = priv
    worker_crypto._job_slug = job_slug
    worker_crypto._post_job_dek = lambda app_url, slug: {
        "serverPub": vector["server_pub"],
        "parcel": vector["parcel"],
    }
    os.environ["NEST_APP_URL"] = "http://interop.local"
    opened = worker_crypto.get_dek(None, {"slug": job_slug})
    assert opened == dek, "Python could not open the Node-sealed job-dek parcel"
    assert hashlib.sha256(opened).hexdigest() == vector["dek_fingerprint"]
    print("✓ Python opens the Node-sealed job-dek parcel via the real get_dek path")

    # A wrong AAD (different slug) MUST fail the GCM open.
    worker_crypto._job_dek = None
    worker_crypto._job_slug = "other-slug"
    try:
        worker_crypto.get_dek(None, {"slug": "other-slug"})
        raise AssertionError("wrong slug (AAD) was NOT rejected")
    except Exception as e:
        assert not isinstance(e, AssertionError), e
    print("✓ wrong job slug (AAD) rejected")
    worker_crypto._reset_job_state()

    # --- Direction 2: Python seals (server role) for Node to open. ---
    new_dek = os.urandom(32)
    server_priv = ec.generate_private_key(ec.SECP256R1())
    worker_pub_key = ec.EllipticCurvePublicKey.from_encoded_point(
        ec.SECP256R1(), worker_pub
    )
    shared = server_priv.exchange(ec.ECDH(), worker_pub_key)
    transport_key = HKDF(
        algorithm=hashes.SHA256(), length=32, salt=_TRANSPORT_SALT, info=_TRANSPORT_INFO
    ).derive(shared)
    nonce = os.urandom(12)
    ct = AESGCM(transport_key).encrypt(nonce, new_dek, job_slug.encode("utf-8"))
    server_pub = server_priv.public_key().public_bytes(
        serialization.Encoding.X962, serialization.PublicFormat.UncompressedPoint
    )
    with open(PY_VECTOR, "w", encoding="utf-8") as fh:
        json.dump(
            {
                "job_slug": job_slug,
                "dek": base64.b64encode(new_dek).decode("ascii"),
                "server_pub": base64.b64encode(server_pub).decode("ascii"),
                "parcel": base64.b64encode(nonce + ct).decode("ascii"),
            },
            fh,
            indent=2,
        )
    print("✓ vector-jobdek-py.json written — re-run verify_jobdek.mjs to close the loop")


if __name__ == "__main__":
    main()
